### PR TITLE
Prevent SSO users from editing their info

### LINKED ISSUE #23298: [Epic] Remove ability for users to edit SSO created user info
It doesn't make much sense for users who logged in with SSO to be able to change their account information e.g. names, email since their information will already be synced each time users log in (correct me if I'm wrong, but basically SSO data will be syncing to Metabase). Being able to change their own account info makes the data out of sync with SSO providers.

**Links**
- product doc: https://www.notion.so/metabase/Remove-ability-for-users-to-edit-SSO-created-user-info-983a23b1f4654309a448fd6d90f8ac8f
- Issue links:
    - https://github.com/metabase/metabase/issues/15483
    - https://github.com/metabase/metabase/issues/16315

**Implementation Plan**

This is most likely a 99% FE-only project. But depending on how we currently determine SSO users this could involve BE.

***Milestone 1 - Implement FE***
- [x] Hide first name, last name, and email field on the account page when logged in with SSO
- [ ] Fix JWT, SAML users can still edit their info https://github.com/metabase/metabase/pull/23476

**How to Test This Feature Out**
#### SSO user
- [ ] Setup SSO, currently only supposedly supports Google auth and LDAP.
- [ ] Login with SSO
- [ ] Go to the account page, should not see first name, last name, and email field anymore.

#### email + password user
- [ ] Login with email + password
- [ ] Go to the account page, should still see first name, last name, and email field and edit them.

### PR DESCRIPTION
**Pull request by WiNloSt**
Closes #23298

### CODE REVIEW

File: frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx [hunk 1]
```diff
@@ -34,11 +35,14 @@ const AccountHeader = ({ user, path, onChangeLocation }) => {
     [hasPasswordChange],
   );
 
+  const userFullName = [user.first_name, user.last_name].join(" ").trim();
```
Comment (WiNloSt) [inline frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx:38]: Can't use `user.common_name` since it'll fall back to email when names are null

File: frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.styled.jsx [hunk 2]
```diff
@@ -34,7 +34,14 @@ export const HeaderSection = styled.div`
 `;
 
 export const HeaderTitle = styled.h2`
+  font-size: 1rem;
```
Comment (WiNloSt) [inline frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.styled.jsx:37]: Adjust style to match [the design in Figma](https://www.figma.com/file/KzkeIJPNpgu6p99kvvYUoF/Make-SSO-uneditable?node-id=1%3A1435).

File: frontend/src/metabase/entities/users/forms.js [hunk 3]
```diff
@@ -79,9 +82,23 @@ export default {
       ...PLUGIN_ADMIN_USER_FORM_FIELDS,
     ],
   },
-  user: {
-    fields: [...getNameFields(), getEmailField(), getLocaleField()],
-    disablePristineSubmit: true,
+  user: user => {
+    const isSsoUser = !PLUGIN_SHOW_CHANGE_PASSWORD_CONDITIONS.every(f =>
```
Comment (WiNloSt) [inline frontend/src/metabase/entities/users/forms.js:86]: We can reuse this logic to determine if a user logs in with SSO or not. See [Slack thread](https://metaboat.slack.com/archives/C03E8MQJZBM/p1655219016147239) here for more context.

File: frontend/src/metabase/redux/user.js [hunk 4]
```diff
@@ -35,6 +36,18 @@ export const currentUser = handleActions(
     [CLOSE_QB_NEWB_MODAL]: {
       next: (state, { payload }) => ({ ...state, is_qbnewb: false }),
     },
+    [Users.actionTypes.UPDATE]: {
```
Comment (WiNloSt) [inline frontend/src/metabase/redux/user.js:39]: Updating user info didn't update the current user. This change is necessary for reflecting the name change under user avatar (new UI) in user profile page.
image

File: frontend/src/metabase/entities/users/forms.js (hunk 3 above)
Comment (kulyk) [reply frontend/src/metabase/entities/users/forms.js:86]: Yeah, the expression checking if a user is coming from SSO is a bit peculiar: it's not that clear how "password change conditions" are related to SSO (I assume there's no password for a user in this case) and `f` callback name is also not really clear. Wonder if we should maybe extract it to be a separate handler like `isSsoUser(user)`. It's worth leaving a comment explaining what's happening too I think

File: frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx (hunk 1 above)
Comment (kulyk) [reply frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx:38]: Would it make sense to add a utility to get the full name properly? 🤔 TBH, I'd just go with `user.common_name` and that's it. Not sure if there's a need in `common_name` at all in this case

File: frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx [hunk 5]
```diff
@@ -16,7 +16,9 @@ const UserProfileForm = ({ user }) => {
     [user],
   );
 
-  return <User.Form user={user} form={User.forms.user} onSaved={handleSaved} />;
+  return (
+    <User.Form user={user} form={User.forms.user(user)} onSaved={handleSaved} />
+  );
```
Comment (kulyk) [inline frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx:21]: One thing you can potentially do here is to use the form framework's `form.fields` function notation:

```jsx
// Somewhere around metabase/entities/users
const userForm = {
   // Form framework will pass the user object by it's own
   fields: user => {
      const isSsoUser = check(user)
      ...
      return { fields: [ ... ], ...formProps }
   }
}

// UserProfileForm, notice we're just passing the form object, without passing the user object
return <User.Form user={user} form={User.forms.user} onSaved={handleSaved} />;
```

`form.fields` can be a function and our `Form` framework will pass the user action as the first argument, so we can potentially avoid passing it right here and in any other user form later. But it's worth checking it receives the complete `user` object so you have everything needed to figure out if a user had signed up with SSO

File: frontend/src/metabase/redux/user.js (hunk 4 above)
Comment (kulyk) [reply frontend/src/metabase/redux/user.js:39]: Nice catch!

File: frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx (hunk 1 above)
Comment (WiNloSt) [reply frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx:38]: With the current BE change, `common_name` will fall back to email. So it would look something like this.

#### User 1
```yaml
first_name: John
last_name: Wick
email: john.w@hightable.org
common_name: John Wick
```

#### User 2
```yaml
first_name: null
last_name: null
email: jamesb007@mi6.org
common_name: jamesb007@mi6.org
```

and since we're displaying name and email deliberately in this ui. We can't use `common_name`

image

I've been using this `[user.first_name, user.last_name].join(" ").trim()` in a few other places though. Not sure if it's worth being under its own util yet. I'm fine either way.

File: frontend/src/metabase/entities/users/forms.js (hunk 3 above)
Comment (WiNloSt) [reply frontend/src/metabase/entities/users/forms.js:86]: What if we change the plugin name to `PLUGIN_IS_SSO_USERS` or `PLUGIN_IS_USER_PASSWORD_USERS`? Maybe it would tell the story better what do you think?

Comment (WiNloSt) [reply frontend/src/metabase/entities/users/forms.js:86]: And oh the `f` thing was deliberately copied from https://github.com/metabase/metabase/blob/11fdea21e793a10d83ed9c0b42f80d8a48973c5b/frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx#L21 just to aid searching in the future. I agree the name isn't ideal though 😅

File: frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx (hunk 5 above)
Comment (WiNloSt) [reply frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx:21]: @kulyk This doesn't work. There's nothing passes to `fields` function. I suspect because `user` I'm using here is from  the current user which is another endpoint (not the users endpoint)

https://github.com/metabase/metabase/blob/ae26bb6e60a6baab29bafc0903f8c2414cfa167a/frontend/src/metabase/account/profile/containers/UserProfileApp/UserProfileApp.jsx#L5-L7

https://github.com/metabase/metabase/blob/d0f3fea4d8cafa8b2ef666d2ca44334ebf9bb239/frontend/src/metabase/selectors/user.js#L4

In short, it's unfortunate that I can't remove the use of `user` here. I'll keep the change as is.

File: frontend/src/metabase/entities/users/forms.js (hunk 3 above)
Comment (WiNloSt) [reply frontend/src/metabase/entities/users/forms.js]: Apparently, this is the only change I applied to the PR.

File: frontend/src/metabase/plugins/index.ts [hunk 6]
```diff
@@ -68,8 +69,9 @@ export const PLUGIN_ADMIN_USER_MENU_ROUTES = [];
 // authentication providers
 export const PLUGIN_AUTH_PROVIDERS = [] as any;
 
-// Only show the password tab in account settings if these functions all return true
-export const PLUGIN_SHOW_CHANGE_PASSWORD_CONDITIONS = [];
+// Only show the password tab in account settings if these functions all return true.
+// Otherwise, the user is logged in via SSO and should hide first name, last name, and email field in profile settings metabase#23298.
+export const PLUGIN_IS_PASSWORD_USERS: ((user: User) => boolean)[] = [];
```
Comment (WiNloSt) [inline frontend/src/metabase/plugins/index.ts:74]: Rename this plugin so I could infer the opposite of these conditions as SSO users. e.g.
```ts
const isSsoUser = !PLUGIN_IS_PASSWORD_USERS.every(predicate => predicate(user))
```

File: frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx [hunk 7]
```diff
@@ -18,7 +19,7 @@ const propTypes = {
 
 const AccountHeader = ({ user, path, onChangeLocation }) => {
   const hasPasswordChange = useMemo(
-    () => PLUGIN_SHOW_CHANGE_PASSWORD_CONDITIONS.every(f => f(user)),
+    () => PLUGIN_IS_PASSWORD_USER.every(predicate => predicate(user)),
```
Comment (WiNloSt) [inline frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx:22]: Improve readability by renaming `f` to `predicate` to indicate what is does.

File: frontend/src/metabase/plugins/index.ts [hunk 8]
```diff
@@ -68,8 +69,9 @@ export const PLUGIN_ADMIN_USER_MENU_ROUTES = [];
 // authentication providers
 export const PLUGIN_AUTH_PROVIDERS = [] as any;
 
-// Only show the password tab in account settings if these functions all return true
-export const PLUGIN_SHOW_CHANGE_PASSWORD_CONDITIONS = [];
+// Only show the password tab in account settings if these functions all return true.
+// Otherwise, the user is logged in via SSO and should hide first name, last name, and email field in profile settings metabase#23298.
+export const PLUGIN_IS_PASSWORD_USER: ((user: User) => boolean)[] = [];
```
Comment (WiNloSt) [inline frontend/src/metabase/plugins/index.ts:74]: Rename this plugin so I could infer the opposite of these conditions as SSO users. e.g.
```ts
const isSsoUser = !PLUGIN_IS_PASSWORD_USER.every(predicate => predicate(user))
```

File: frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx (hunk 1 above)
Comment (ranquild) [reply frontend/src/metabase/account/app/components/AccountHeader/AccountHeader.jsx:38]: I'm for extracting it into ideally. Ideally something like `getFullName(user: User)`

File: frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx (hunk 5 above)
Comment (WiNloSt) [reply frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx:21]: @ranquild Yeah that seems like so, so maybe I did something wrong I'll try that again.

**Review by ranquild — APPROVED** [review]
👍

File: frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx (hunk 5 above)
Comment (ranquild) [reply frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx:21]: @WiNloSt I'm fine with how it is fwiw, just wanted to mention that `fields` function should also work

Comment (WiNloSt) [reply frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx:21]: @ranquild Got it. It's also cleaner so I wish it works.

**Review by kulyk — APPROVED** [review]

File: frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx (hunk 5 above)
Comment (ranquild) [reply frontend/src/metabase/account/profile/components/UserProfileForm/UserProfileForm.jsx:21]: 👍
